text/0000-impl-specialization.md

```diff
@@ -0,0 +1,1652 @@
+- Feature Name: specialization
+- Start Date: 2015-06-17
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+This RFC proposes a design for *specialization*, which permits multiple `impl`
+blocks to apply to the same type/trait, so long as one of the blocks is clearly
+"more specific" than the other. The more specific `impl` block is used in a case
+of overlap. The design proposed here also supports refining default trait
+implementations based on specifics about the types involved.
+
+Altogether, this relatively small extension to the trait system yields benefits
+for performance and code reuse, and it lays the groundwork for an "efficient
+inheritance" scheme that is largely based on the trait system (described in a
+forthcoming companion RFC).
+
+# Motivation
+
+Specialization brings benefits along several different axes:
+
+* **Performance**: specialization expands the scope of "zero cost abstraction",
+  because specialized impls can provide custom high-performance code for
+  particular, concrete cases of an abstraction.
+
+* **Reuse**: the design proposed here also supports refining default (but
+  incomplete) implementations of a trait, given details about the types
+  involved.
+
+* **Groundwork**: the design lays the groundwork for supporting
+  ["efficient inheritance"](https://internals.rust-lang.org/t/summary-of-efficient-inheritance-rfcs/494)
+  through the trait system.
+
+The following subsections dive into each of these motivations in more detail.
+
+## Performance
+
+The simplest and most longstanding motivation for specialization is
+performance.
+
+To take a very simple example, suppose we add a trait for overloading the `+=`
+operator:
+
+```rust
+trait AddAssign<Rhs=Self> {
+    fn add_assign(&mut self, Rhs);
+}
+```
+
+It's tempting to provide an impl for any type that you can both `Clone` and
+`Add`:
+
+```rust
+impl<R, T: Add<R> + Clone> AddAssign<R> for T {
+    fn add_assign(&mut self, rhs: R) {
+        let tmp = self.clone() + rhs;
+        *self = tmp;
+    }
+}
+```
+
+This impl is especially nice because it means that you frequently don't have to
+bound separately by `Add` and `AddAssign`; often `Add` is enough to give you
+both operators.
+
+However, in today's Rust, such an impl would rule out any more specialized
```

Comment:
I think it would help readability if there was an example of a desired impl that couldn't coeexist with the blanket impl today here.